src/antsibull_core/collection_meta.py

```diff
@@ -46,6 +48,95 @@ def __init__(self, *, all_collections: list[str], major_release: int):
+        field_name: str,
+    ) -> str:
+        if state not in accepted_states:
+            if state is not None:
+                self.errors.append(
+                    f"{prefix}[{index}] -> {field_name}: Unexpected update after {state}"
+                )
+            else:
+                self.errors.append(
+                    f"{prefix}[{index}] -> {field_name}: Unexpected first update"
+                )
+        return field_name
+
+    def _update_state(
+        self, state: str | None, index: int, update: RemovalUpdate, prefix: str
+    ) -> tuple[str | None, PypiVer | None, str]:
+        if update.cancelled_version:
+            state = self._update_state_value(
+                state,
+                [None, "deprecated_version", "redeprecated_version"],
+                prefix,
+                index,
+                "cancelled_version",
+            )
+            return state, update.cancelled_version, "cancelled_version"
+        if update.deprecated_version:
+            state = self._update_state_value(
+                state, [None], prefix, index, "deprecated_version"
+            )
+            return state, update.deprecated_version, "deprecated_version"
+        if update.redeprecated_version:
+            state = self._update_state_value(
+                state,
+                ["readded_version", "cancelled_version"],
+                prefix,
+                index,
+                "redeprecated_version",
+            )
+            return state, update.redeprecated_version, "redeprecated_version"
+        if update.removed_version:
+            state = self._update_state_value(
+                state, [None], prefix, index, "removed_version"
+            )
+            return state, update.removed_version, "removed_version"
+        if update.readded_version:
+            state = self._update_state_value(
+                state, ["removed_version"], prefix, index, "readded_version"
+            )
+            return state, update.readded_version, "readded_version"
+        # The following lines should never be reached:
+        self.errors.append(f"{prefix}[{index}]: Internal error")  # pragma: no cover
+        return state, None, ""  # pragma: no cover
+
+    def _validate_removal_updates(
+        self,
+        removal: BaseRemovalInformation,
+        indirect_updates: list[RemovalUpdate],
+        prefix: str,
+    ) -> None:
+        prefix += " -> updates"
+        state = None
+        for update in indirect_updates:
+            state, _, __ = self._update_state(state, -1, update, prefix)
+        last_version = None
+        for index, update in enumerate(removal.updates):
+            state, version, field_name = self._update_state(
+                state, index, update, prefix
+            )
+            if version is None:
+                # The following line should never be reached:
+                pass  # pragma: no cover
+            elif version.major != self.major_release:
```

Comment:
No, we'd keep the `removal` metadata, but copy it back from `removed_collections` to `collections` and add `updates:` and move `announce_version` to `removed_version` there, and re-add `major_version`. https://github.com/ansible-community/ansible-build-data/pull/488/files#diff-de27baf579d836d439eb1aa79999d825ae044f35799bb2e89b598bde6791a047 shows how it looks like.

Reply:
For the `netapp.storagegrid` example where the collection was re-added, I'm asking about what would happen when we create the `12` directory. We would just remove its `removal:` metadata because it would no longer apply to Ansible 12, right?

Reply:
Yes. If it later happens to be deprecated again in Ansible 12, a new `removal` entry can be added.